lib/pgBackRest/Storage/S3/Request.pm

```diff
@@ -0,0 +1,202 @@
+{
+    my $class = shift;
+
+    # Create the class hash
+    my $self = {};
+    bless $self, $class;
+
+    # Assign function parameters, defaults, and log debug info
+    (
+        my $strOperation,
+        $self->{strBucket},
+        $self->{strEndPoint},
+        $self->{strRegion},
+        $self->{strAccessKeyId},
+        $self->{strSecretAccessKey},
+        $self->{strHost},
+        $self->{bVerifySsl},
+        $self->{lBufferMax},
+    ) =
+        logDebugParam
+        (
+            __PACKAGE__ . '->new', \@_,
+            {name => 'strBucket', trace => true},
+            {name => 'strEndPoint', trace => true},
+            {name => 'strRegion', trace => true},
+            {name => 'strAccessKeyId', trace => true},
+            {name => 'strSecretAccessKey', trace => true},
+            {name => 'strHost', optional => true, trace => true},
+            {name => 'bVerifySsl', optional => true, default => true, trace => true},
+            {name => 'lBufferMax', optional => true, default => COMMON_IO_BUFFER_MAX, trace => true},
+        );
+
+    # If host is not set then it will be bucket + endpoint
+    $self->{strHost} = defined($self->{strHost}) ? $self->{strHost} : "$self->{strBucket}.$self->{strEndPoint}";
+
+    # Return from function and log return values if any
+    return logDebugReturn
+    (
+        $strOperation,
+        {name => 'self', value => $self, trace => true}
+    );
+}
+
+####################################################################################################################################
+# request - send a request to S3
+####################################################################################################################################
+sub request
+{
+    my $self = shift;
+
+    # Assign function parameters, defaults, and log debug info
+    my
+    (
+        $strOperation,
+        $strVerb,
+        $strUri,
+        $hQuery,
+        $hHeader,
+        $rstrBody,
+        $strResponseType,
+        $bIgnoreMissing,
+    ) =
+        logDebugParam
+        (
+            __PACKAGE__ . '->request', \@_,
+            {name => 'strVerb', trace => true},
+            {name => 'strUri', optional => true, default => '/', trace => true},
+            {name => 'hQuery', optional => true, trace => true},
+            {name => 'hHeader', optional => true, trace => true},
+            {name => 'rstrBody', optional => true, trace => true},
+            {name => 'strResponseType', optional => true, default => S3_RESPONSE_TYPE_NONE, trace => true},
+            {name => 'bIgnoreMissing', optional => true, default => false, trace => true},
+        );
+
+    # Get datetime to be used for auth requests
+    my $strDateTime = s3DateTime();
+
+    # Set content length and hash
+    $hHeader->{&S3_HEADER_CONTENT_SHA256} = defined($rstrBody) ? sha256_hex($$rstrBody) : PAYLOAD_DEFAULT_HASH;
+    $hHeader->{&S3_HEADER_CONTENT_LENGTH} = defined($rstrBody) ? length($$rstrBody) : 0;
```

Comment:
I read that to mean you should send it when there is a message.  If there's no message there is no need.

It's definitely not calculated in HttpClient - and can't be since content-length if present must be part of Signed Headers.